File: modules/waste_recognition/src/store.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

// Vendored per-module datastore. Intentionally NOT shared with sibling
// modules — a buyer of this directory gets a working store with no
// cross-repo dependency to untangle.
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DB_FILE = path.join(__dirname, "data", "store.json");

const EMPTY = { classifications: [] };

function load() {
  if (!fs.existsSync(DB_FILE)) {
    fs.mkdirSync(path.dirname(DB_FILE), { recursive: true });
    fs.writeFileSync(DB_FILE, JSON.stringify(EMPTY, null, 2));
  }
  return JSON.parse(fs.readFileSync(DB_FILE, "utf-8"));
}

export const store = {
  read: load,
  write(data) {
    fs.writeFileSync(DB_FILE, JSON.stringify(data, null, 2));
  },
};

File: hacquire/modules/route_optimizer/app/store.py
"""Module-local JSON datastore.

VENDORED, NOT SHARED. Every module carries its own copy of this file. That
duplicates ~40 lines across the registry — the deliberate price of being able
to hand any single directory to a buyer and have it run with no cross-repo
dependency to untangle.

Swap the class body for a real database client and no route changes.
"""
import json
import os
import secrets
import threading
from datetime import datetime, timezone
from pathlib import Path


class Store:
    def __init__(self, empty: dict, filename: str = "store.json"):
        self._empty = empty
        self._path = Path(__file__).resolve().parent / "data" / filename
        self._lock = threading.Lock()
        self._path.parent.mkdir(parents=True, exist_ok=True)
        if not self._path.exists():
            self._write_unlocked(empty)

    def _write_unlocked(self, data: dict) -> None:
        # Write-then-rename: a crash mid-write leaves the previous file intact
        # rather than a truncated one.
        tmp = self._path.with_suffix(".tmp")
        tmp.write_text(json.dumps(data, indent=2))
        tmp.replace(self._path)

    def read(self) -> dict:
        with self._lock:
            try:
                return json.loads(self._path.read_text())
            except (FileNotFoundError, json.JSONDecodeError):
                return json.loads(json.dumps(self._empty))

    def write(self, data: dict) -> None:
        with self._lock:
            self._write_unlocked(data)

    def reset(self) -> None:
        self.write(json.loads(json.dumps(self._empty)))

    @staticmethod
    def new_id(prefix: str) -> str:
        return f"{prefix}_{secrets.token_hex(6)}"

    @staticmethod
    def now() -> str:
        return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
